proxyclient/client.go

```
@@ -0,0 +1,186 @@
+
+var (
+	nonTLSDialer = &websocket.Dialer{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+	}
+	retryTimeout = 1 * time.Second
+)
+
+type PortForwarder interface {
+	Start() error
+	Stop()
+}
+
+type ProxyClientOpt func(*ProxyClient)
+
+type ProxyClient struct {
+	forwarder           PortForwarder
+	serverUrl           string
+	serverConnectSecret string
+	dialer              *websocket.Dialer
+	secretController    v1.SecretController
+
+	onConnect func(ctx context.Context, session *remotedialer.Session) error
+}
+
+func New(serverSharedSecret, namespace, certSecretName, certServerName string, restConfig *rest.Config, forwarder PortForwarder, opts ...ProxyClientOpt) (*ProxyClient, error) {
+	if restConfig == nil {
+		return nil, fmt.Errorf("restConfig required")
+	}
+
+	if forwarder == nil {
+		return nil, fmt.Errorf("a PortForwarder must be provided")
+	}
+
+	if namespace == "" {
+		return nil, fmt.Errorf("namespace required")
+	}
+
+	if certSecretName == "" {
+		return nil, fmt.Errorf("certSecretName required")
+	}
+
+	if serverSharedSecret == "" {
+		return nil, fmt.Errorf("server shared secret must be provided")
+	}
+
+	serverUrl := fmt.Sprintf("%s:%d%s", defaultServerAddr, defaultServerPort, defaultServerPath)
+	dialer, err := buildDialer(namespace, certSecretName, certServerName, restConfig)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't build dialer: %w", err)
+	}
+
+	client := &ProxyClient{
+		serverUrl:           serverUrl,
+		forwarder:           forwarder,
+		dialer:              dialer,
+		serverConnectSecret: serverSharedSecret,
+	}
+
+	for _, opt := range opts {
+		opt(client)
+	}
+
+	return client, nil
+}
+
+func buildDialer(namespace, certSecretName, certServerName string, restConfig *rest.Config) (*websocket.Dialer, error) {
+	core, err := core.NewFactoryFromConfigWithOptions(restConfig, nil)
+	if err != nil {
+		logrus.Error("build secret controller failed: %w, defaulting to non TLS connection", err)
+		return nonTLSDialer, err
```

This makes me a bit nervous, is there a reason for falling back to insecure here?

---

Not specifically. Do you prefer I just return the err and fail it ?

---

Yes let's just return an error and fail. We don't need to support non-TLS connections (at least for now).